src/protocol.errorcode.ts

```diff
@@ -3,7 +3,21 @@
  */
 export enum ProtocolErrorCode {
     COMMON_ERROR = 'CommonError',
-    FINGERPRINT_NOMATCH = 'FingerprintNoMatch',
+    BIOANALYZER_SERVER_ERROR = 'BioanalyzerServerError',
```

Comment:
I just realized something with these.   These errors are specific to kiva, and only deployments that use biometrics.  Some consumers (maybe not kiva) would never need these codes.  I am wondering if e should make a different enum here.   If we do, what complications if any does that create?  thoughts?

Reply:
Perhaps it makes more sense to put them in a `BioAuthErrorCode` enum? It doesn't really create any additional complications since this is all new to begin with (except for FINGERPRINT_NO_MATCH, but that's something I was going to change as part of this anyway).

I think it still makes sense to have them be available somewhere, because the whole system is open source and I think some people may indeed be interested in using those biometrics pieces. Not to mention that keeping these cross-service communication pieces in a centralized, common location makes maintenance of the system much easier.

I see what you mean about it being not part of the core agent-management loop, since people may have different sorts of authentication mechanisms, but I think it's maybe a bit too strong to say that only kiva would ever use them.

Reply:
On one hand there's a risk that we keep a list of all possible errors types of all implementations which could become unwieldy to maintain and confusing to readers. On the other hand there's a risk that by breaking it up or making it more abstract we complicate our error handling and that causes debugging to be harder. Both seem like problems. I suppose the severity of the problem depends on how many additional error types we end up adding over the next few years. I don't have strong opinions only wanted to highlight the tradeoff. My slight preference would be sticking with one big list for now

Reply:
Just pushed a couple updates that show what it would be like to have several collections of error codes instead of just having one big list.

Reply:
I'm usually all about modularization and I do think there's some organizational benefit in this case. The thing I worry about is when people start trying to handle all the possible error types in, for example, a top-level error handler. It could end up with one of those painfully boilerplate-y try-catch waterfalls, or people could be tempted to abandon type-casting and start using `any` or `string` because it's easier than trying to maintain a union type of all the possible error codes. We could add a union type ourselves, but then we'd have to maintain it and, frankly, using it abandons a lot of the things that are good about enums to begin with.

Reply:
Update: Offline discussion resulted in us deciding to keep all the error codes in one enum until we have a better idea of what the abstractions should be.